x-pack/auditbeat/processors/add_session_metadata/processdb/db.go

```diff
@@ -0,0 +1,739 @@
+// situation to return an empty string
+func basename(pathStr string) string {
+	if pathStr == "" {
+		return ""
+	}
+
+	return path.Base(pathStr)
+}
+
+func (db *DB) InsertFork(fork types.ProcessForkEvent) error {
+	db.Lock()
+	defer db.Unlock()
+
+	pid := fork.ChildPids.Tgid
+	ppid := fork.ParentPids.Tgid
+	if entry, ok := db.processes[ppid]; ok {
+		entry.Pids = pidInfoFromProto(fork.ChildPids)
+		entry.Creds = credInfoFromProto(fork.Creds)
+		db.processes[pid] = entry
+		if entryPid, ok := db.entryLeaderRelationships[ppid]; ok {
+			db.entryLeaderRelationships[pid] = entryPid
+		}
+	} else {
+		db.processes[pid] = Process{
+			Pids:  pidInfoFromProto(fork.ChildPids),
+			Creds: credInfoFromProto(fork.Creds),
+		}
+	}
+
+	return nil
+}
+
+func (db *DB) insertProcess(process Process) {
+	pid := process.Pids.Tgid
+	db.processes[pid] = process
+	entryLeaderPid := db.evaluateEntryLeader(process)
+	if entryLeaderPid != nil {
+		db.entryLeaderRelationships[pid] = *entryLeaderPid
+		db.logger.Debugf("%v name: %s, entry_leader: %d, entry_type: %s", process.Pids, process.Filename, *entryLeaderPid, string(db.entryLeaders[*entryLeaderPid]))
+	} else {
+		db.logger.Debugf("%v name: %s, NO ENTRY LEADER", process.Pids, process.Filename)
+	}
+}
+
+func (db *DB) InsertExec(exec types.ProcessExecEvent) error {
+	db.Lock()
+	defer db.Unlock()
+
+	proc := Process{
+		Pids:             pidInfoFromProto(exec.Pids),
+		Creds:            credInfoFromProto(exec.Creds),
+		CTty:             ttyDevFromProto(exec.CTty),
+		Argv:             exec.Argv,
+		Cwd:              exec.Cwd,
+		Env:              exec.Env,
+		Filename:         exec.Filename,
+	}
+
+	db.processes[exec.Pids.Tgid] = proc
+	entryLeaderPid := db.evaluateEntryLeader(proc)
+	if entryLeaderPid != nil {
+		db.entryLeaderRelationships[exec.Pids.Tgid] = *entryLeaderPid
+	}
+
+	return nil
+}
+
+func (db *DB) createEntryLeader(pid uint32, entryType EntryType) {
+	db.entryLeaders[pid] = entryType
+	db.logger.Debugf("created entry leader %d: %s, name: %s", pid, string(entryType), db.processes[pid].Filename)
+}
+
+// pid returned is a pointer type because its possible for no
+func (db *DB) evaluateEntryLeader(p Process) *uint32 {
+	pid := p.Pids.Tgid
+
+	// init never has an entry leader or meta type
+	if p.Pids.Tgid == 1 {
+		db.logger.Debugf("entry_eval %d: process is init, no entry type", p.Pids.Tgid)
+		return nil
+	}
+
+	// kernel threads also never have an entry leader or meta type kthreadd
+	// (always pid 2) is the parent of all kernel threads, by filtering pid ==
+	// 2 || ppid == 2, we get rid of all of them
+	if p.Pids.Tgid == 2 || p.Pids.Ppid == 2 {
+		db.logger.Debugf("entry_eval %d: kernel threads never an entry type (parent is pid 2)", p.Pids.Tgid)
+		return nil
+	}
+
+	// could be an entry leader
+	if p.Pids.Tgid == p.Pids.Sid {
+		ttyType := getTtyType(p.CTty.Major, p.CTty.Minor)
+
+		procBasename := basename(p.Filename)
+		if ttyType == Tty {
```

Comment:
Suggest using a `switch` here.

Reply:
I don't think switch would work well here, each if statement is comparing something different. I don't think it would be any cleaner to use switch with it

Reply:
The reason I make the suggestion is that the conditions are lighter on the page in a switch; compare:

```
		switch {
		case ttyType == Tty:
			db.createEntryLeader(pid, Terminal)
			db.logger.Debugf("entry_eval %d: entry type is terminal", p.Pids.Tgid)
			return &pid

		case ttyType == TtyConsole && procBasename == "login":
			db.createEntryLeader(pid, EntryConsole)
			db.logger.Debugf("entry_eval %d: entry type is console", p.Pids.Tgid)
			return &pid

		case p.Pids.Ppid == 1:
			db.createEntryLeader(pid, Init)
			db.logger.Debugf("entry_eval %d: entry type is init", p.Pids.Tgid)
			return &pid

		case !isFilteredExecutable(procBasename):
			if parent, ok := db.processes[p.Pids.Ppid]; ok {
				parentBasename := basename(parent.Filename)
				switch {
				case ttyType == Pts && parentBasename == "ssm-session-worker":
					db.createEntryLeader(pid, Ssm)
					db.logger.Debugf("entry_eval %d: entry type is ssm", p.Pids.Tgid)
					return &pid

				case parentBasename == "sshd" && procBasename != "sshd":
					// TODO: get ip from env vars
					db.createEntryLeader(pid, Sshd)
					db.logger.Debugf("entry_eval %d: entry type is sshd", p.Pids.Tgid)
					return &pid
					
				case isContainerRuntime(parentBasename):
					db.createEntryLeader(pid, Container)
					db.logger.Debugf("entry_eval %d: entry type is container", p.Pids.Tgid)
					return &pid
				}
			}

		default:
			db.logger.Debugf("entry_eval %d: is a filtered executable: %s", p.Pids.Tgid, procBasename)
		}
```

Reply:
Thanks, I was thinking of a C style switch. I've changed to this